### PR TITLE
feat(drive): implement ABCI method OfferSnapshot

### PR DESCRIPTION
**Pull request by shotonoff**
## Issue being fixed or feature implemented

`OfferSnapshot` ABCI method is called by Tenderdash to offer a snapshot for synchronization. 

## What was done?

* If the offered snapshot is invalid then ABCI returns error
* If the snapshot was offered before, then ABCI has to wipe the synched data from database and accept the snapshot to proceed replication

## How Has This Been Tested?

Local tests

## Breaking Changes

N/A

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive-abci/src/platform_types/snapshot/mod.rs [hunk 1]
```diff
@@ -168,4 +207,32 @@ mod tests {
...
+        let snapshot_2000 = abci::Snapshot {
+            height: 2000,
+            version: 0,
+            hash: app_hash.clone(),
+            metadata: vec![],
+        };
+        manager
+            .offer_snapshot(&grove, snapshot_2000.clone(), app_hash.clone())
+            .unwrap();
+        fs::remove_dir_all(grove_dir.path()).unwrap();
```
Comment (lklimek) [inline packages/rs-drive-abci/src/platform_types/snapshot/mod.rs:236]: The tempdir function creates a directory in the file system and returns a TempDir. The directory will be automatically deleted when the TempDirs destructor is run.

File: packages/rs-drive/Cargo.toml [hunk 2]
```diff
@@ -79,7 +82,7 @@ default = ["full"]
 fixtures-and-mocks = ["mockall"]
 full = [
     "fixtures-and-mocks",
-    "grovedb/full",
+    "grovedb/estimated_costs",
```
Comment (lklimek) [inline packages/rs-drive/Cargo.toml:85]: Why this change?

File: packages/rs-drive-abci/src/platform_types/snapshot/mod.rs [hunk 3]
```diff
@@ -53,6 +68,30 @@ impl Manager {
...
+        grove: &GroveDb,
+        snapshot: abci::Snapshot,
+        app_hash: Vec<u8>,
+    ) -> Result<response_offer_snapshot::Result, Error> {
+        match take(&mut self.offered_snapshot) {
+            Some(mut offered_snapshot) => {
+                if offered_snapshot.snapshot.height == snapshot.height {
+                    return Ok(response_offer_snapshot::Result::Reject);
+                }
+                if offered_snapshot.snapshot.version != snapshot.version {
```
[the remark this answers is not in the document]
Comment (shotonoff) [reply packages/rs-drive-abci/src/platform_types/snapshot/mod.rs:86]: I introduced the version through the constant `SNAPSHOT_VERSION`

**Review by shotonoff — COMMENTED** [review]
The PR is ready for review

File: packages/rs-drive/Cargo.toml (hunk 2 above)
Comment (iammadab) [reply packages/rs-drive/Cargo.toml:96]: Yeah, you should also look at @QuantumExplorer pr: https://github.com/dashpay/platform/pull/1340

File: packages/rs-drive-abci/src/platform_types/snapshot/mod.rs [hunk 4]
```diff
@@ -168,4 +207,32 @@ mod tests {
             fs::remove_dir_all(grove_dir.path()).unwrap();
         }
     }
+
+    #[test]
+    fn test_offer_snapshot() {
+        let grove_dir = tempfile::tempdir().unwrap();
+        let replication_dir = tempfile::tempdir().unwrap();
+        let grove = GroveDb::open(grove_dir.path()).unwrap();
+        let mut manager = Manager::new("".to_string(), Some(3), Some(1));
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-drive-abci/src/platform_types/snapshot/mod.rs:218]: My concern is that this will lead to writing files in a current working dir, which is not what we want. If it doesn't matter now, it can matter in future. 

So, as a good practice, we should initialize correctly, with a tempdir.

**Review by lklimek — COMMENTED** [review]
I see some warnings in `packages/rs-drive-abci/src/platform_types/snapshot/mod.rs`, please fix them

**Review by shotonoff — COMMENTED** [review]
answered

**Review by QuantumExplorer — APPROVED** [review]
